semgrep-core/src/fixing/Autofix_printer.ml

```diff
@@ -0,0 +1,128 @@
+(* Nat Mote
+ *
+ * Copyright (C) 2019-2022 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open Common
+open AST_generic
+module MV = Metavariable
+
+(******************************************************************************)
+(* Handles AST printing for the purposes of autofix.
+ *
+ * The main printing logic happens elsewhere. This module's main purpose is to
+ * extend the existing AST printers so that they can avoid printing AST nodes
+ * which are lifted unchanged from either the target file (via metavariable
+ * bindings) or the fix pattern in the rule. This serves two purposes:
+ * - It lets us synthesize autofixes without having to implement printing for
+ *   all of the AST nodes in the fixed AST.
+ * - It lets us make minimal changes to target files by carrying over
+ *   formatting, comments, etc. from the original code.
+ *)
+(******************************************************************************)
+
+module PythonPrinter = Hybrid_print.Make (struct
+  class printer = Ugly_print_AST.python_printer
+end)
+
+let get_printer lang external_printer : Ugly_print_AST.printer_t option =
+  match lang with
+  | Lang.Python -> Some (new PythonPrinter.printer external_printer)
+  | _ -> None
+
+let original_source_of_ast source any =
+  let* start, end_ = Visitor_AST.range_of_any_opt any in
+  let starti = start.Parse_info.charpos in
+  let _, _, endi = Parse_info.get_token_end_info end_ in
+  let len = endi - starti in
+  let str = String.sub source starti len in
+  Some str
+
+(* Add each metavariable value to the lookup table so that it can be identified
+ * during printing *)
+let add_metavars tbl metavars =
+  List.iter
+    (fun (_, mval) ->
+      let any = MV.mvalue_to_any mval in
+      Hashtbl.replace tbl any `Metavar;
```

Comment:
I'm so confused about what this table looks like 😂. Just going to keep reading and come back

Reply:
Ok, I kind of get it now. We're marking each of the nodes as metavar or pattern, which the pretty-printer will use to decide what ast to print as. This could use explanation in the file header. Still a bit confused about "pattern"

Reply:
Sounds good, I'll explain this better. Maybe I should change Pattern to Fix, since that is for nodes that came from the `fix` field in the rule?